### PR TITLE
Cycles prop tests

### PR DESCRIPTION
**Pull request by bdemann**
- [x] 9999992n counter example for no cycles?
- [x] 4012619n,9999998n counter example for variable
- [x] 1205962n,6372056n counter example for variable
- [x] 4n,6n counter example for variable
- [x] 6n,10n counter example for chunk

### CODE REVIEW

File: tests/property/ic_api/cycles/src/cycles/index.ts
Comment (lastmjs) [inline tests/property/ic_api/cycles/src/cycles/index.ts:1]: Make the accept amount arbitrary

File: tests/property/ic_api/cycles/src/cycles/index.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        //     // TODO we probably don't need to tests this as part of the property tests
+        //     // TODO but if we do should we do this same test but with reject?
+        //     throw new Error(
+        //         `msgCyclesAvailable() !== 0. Actual value: ${remainingCycles}. Initial msgCyclesAvailable result: ${msgCyclesAvailableResult}`
+        //     );
+        // }
+    }
+}
+
+function acceptCycles(
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/src/cycles/index.ts:48]: Pass in the amount to divide by, make that amount arbitrary

File: tests/property/ic_api/cycles/src/cycles/index.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import { CyclesResult, CyclesResultIDL } from '../types';
+
+export default class {
+    @update([], CyclesResultIDL)
+    receiveAllCycles(): CyclesResult {
+        return acceptCycles((available) => available);
+    }
+
+    @update([], CyclesResultIDL)
+    receiveHalfCycles(): CyclesResult {
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/src/cycles/index.ts:19]: Make this arbitrary

File: tests/property/ic_api/cycles/src/cycles/index.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        return acceptCycles((available) => available / 2n);
+    }
+
+    @update([], CyclesResultIDL)
+    receiveNoCycles(): CyclesResult {
+        return acceptCycles(() => 0n);
+    }
+
+    // TODO we could have fun playing with the constraints of msgCyclesAccept
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/src/cycles/index.ts:28]: Consider how making the example more arbitrary like we have discussed would solve these comments

File: tests/property/ic_api/cycles/src/cycles/index.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    // TODO like making sure it doesn't trap, or that it can be called multiple times
+    // TODO or what happens if it accepts more than is available, or it it tries to accept more than the max_amount
+
+    @update([], IDL.Nat64, { manual: true })
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/src/cycles/index.ts:32]: Let's just remove this

File: tests/property/ic_api/cycles/src/types.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+export type CyclesResult = {
+    initialAvailable: bigint;
+    accepted: bigint;
+    finalAvailable: bigint;
+    startingCanisterBalance: bigint;
+    endingCanisterBalance: bigint;
+    cyclesRefunded: bigint;
+};
+
+export const CyclesResultIDL = IDL.Record({
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/src/types.ts:12]: Rename to CyclesResult

File: tests/property/ic_api/cycles/src/intermediary/index.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            cycles: amount
+        });
+        return result;
+    }
+}
+
+function getCyclesPrincipal(): string {
+    return (
+        process.env.CYCLES_PRINCIPAL ??
+        trap('process.env.CYCLES_PRINCIPAL is undefined')
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/src/intermediary/index.ts:50]: throw

File: tests/property/ic_api/cycles/src/intermediary/index.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        );
+        const result = await call(getCyclesPrincipal(), 'receiveCyclesManual', {
+            returnIdlType: IDL.Nat64,
+            cycles: amount
+        });
+        return result;
+    }
+}
+
+function getCyclesPrincipal(): string {
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/src/intermediary/index.ts:47]: Have a class instance variable initialized by calling this function

File: tests/property/ic_api/cycles/src/intermediary/index.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    async sendNoCycles(amount: bigint): Promise<CyclesResult> {
+        const result = await call(getCyclesPrincipal(), 'receiveNoCycles', {
+            returnIdlType: CyclesResultIDL,
+            cycles: amount
+        });
+        return { ...result, cyclesRefunded: msgCyclesRefunded() };
+    }
+
+    @update([IDL.Nat64], IDL.Nat64)
+    async sendCyclesManual(amount: bigint): Promise<bigint> {
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/src/intermediary/index.ts:34]: Get rid of

File: tests/property/ic_api/cycles/test/pretest.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import { getCanisterId } from 'azle/dfx';
+import { linkAndInstallPatch } from 'azle/test/jest_link';
+import { execSync } from 'child_process';
+import { join } from 'path';
+
+function pretest(): void {
+    linkAndInstallPatch(
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/test/pretest.ts:7]: Remove this

File: tests/property/ic_api/cycles/test/pretest.ts
Comment (lastmjs) [inline tests/property/ic_api/cycles/test/pretest.ts:1]: Make this right

File: tests/property/ic_api/cycles/test/test.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import { getCanisterId } from 'azle/dfx';
+import { runTests } from 'azle/test';
+
+import { createActor } from './dfx_generated/intermediary';
+import { getTests } from './tests';
+
+const intermediaryCanister = createActor(getCanisterId('intermediary'), {
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/test/test.ts:7]: Always use that one function thing called something like getCanisterActor

Comment (lastmjs) [reply tests/property/ic_api/cycles/test/test.ts:7]: See if the top-level await will work here

File: tests/property/ic_api/cycles/tsconfig.json
Comment (lastmjs) [inline tests/property/ic_api/cycles/tsconfig.json:1]: Enable top-level await with target and module 2022...

File: tests/property/ic_api/cycles/test/tests.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
+import { ActorSubclass } from '@dfinity/agent';
+import { defaultPropTestParams, expect, it, Test } from 'azle/test';
+import fc from 'fast-check';
+
+import { CyclesResult } from '../src/types';
+// @ts-ignore this path may not exist when these tests are imported into other test projects
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/test/tests.ts:6]: Let's remove unless needed

File: tests/property/ic_api/cycles/test/tests.ts
Comment (lastmjs) [inline tests/property/ic_api/cycles/test/tests.ts:1]: send 0n for all

File: tests/property/ic_api/cycles/test/tests.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+                fc.asyncProperty(fc.bigInt(1n, 10_000_000n), async (amount) => {
+                    const result =
+                        await intermediaryCanister.sendNoCycles(amount);
+                    validateCyclesResult(result, amount, 'none');
+                }),
+                defaultPropTestParams
+            );
+        });
+
+        it('should handle sendCyclesManual correctly', async () => {
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/test/tests.ts:44]: Remove this

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: tests/property/ic_api/cycles/test/tests.ts (hunk 12 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/test/tests.ts:44]: done

File: tests/property/ic_api/cycles/test/tests.ts
Comment (bdemann) [reply tests/property/ic_api/cycles/test/tests.ts:1]: I don't remember what this means

File: tests/property/ic_api/cycles/tsconfig.json
Comment (bdemann) [reply tests/property/ic_api/cycles/tsconfig.json:1]: didn't work because of jest

File: tests/property/ic_api/cycles/test/test.ts (hunk 10 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/test/test.ts:7]: it doesn't because of jest

File: tests/property/ic_api/cycles/test/pretest.ts (hunk 9 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/test/pretest.ts:7]: done

File: tests/property/ic_api/cycles/src/types.ts (hunk 5 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/src/types.ts:12]: done

File: tests/property/ic_api/cycles/src/intermediary/index.ts (hunk 6 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/src/intermediary/index.ts:50]: done

File: tests/property/ic_api/cycles/src/intermediary/index.ts (hunk 7 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/src/intermediary/index.ts:47]: done

File: tests/property/ic_api/cycles/src/intermediary/index.ts (hunk 8 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/src/intermediary/index.ts:34]: done

File: tests/property/ic_api/cycles/src/cycles/index.ts (hunk 1 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/src/cycles/index.ts:48]: done

File: tests/property/ic_api/cycles/src/cycles/index.ts (hunk 4 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/src/cycles/index.ts:32]: done

File: tests/property/ic_api/cycles/src/cycles/index.ts
Comment (bdemann) [reply tests/property/ic_api/cycles/src/cycles/index.ts:1]: done

File: tests/property/ic_api/cycles/src/cycles/index.ts (hunk 2 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/src/cycles/index.ts:19]: done

File: tests/property/ic_api/cycles/src/cycles/index.ts (hunk 3 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/src/cycles/index.ts:28]: done

File: tests/property/ic_api/cycles/test/tests.ts
Comment (bdemann) [reply tests/property/ic_api/cycles/test/tests.ts:1]: I remembered, done

File: tests/property/ic_api/cycles/test/tests.ts [hunk 13]
```diff
@@ -58,16 +83,22 @@ export function getTests(intermediaryCanister: ActorSubclass<_SERVICE>): Test {
...
     if (mode === 'all') {
         expect(result.accepted).toBe(result.initialAvailable);
     }
-    if (mode === 'half') {
-        expect(result.accepted).toBe(result.initialAvailable / 2n);
+    if (mode === 'variable') {
+        if (amountToAccept === undefined) {
+            throw new Error('amountToAccept is required for variable mode');
+        }
+        const accepted =
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/test/tests.ts:99]: Can we review this logic?

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: tests/property/ic_api/cycles/test/tests.ts [hunk 14]
```diff
@@ -1,16 +1,17 @@
-import { ActorSubclass } from '@dfinity/agent';
+import { getCanisterActor } from 'azle/test';
 import { defaultPropTestParams, expect, it, Test } from 'azle/test';
 import fc from 'fast-check';
 
 import { CyclesResult } from '../src/types';
-// @ts-ignore this path may not exist when these tests are imported into other test projects
 import { _SERVICE } from './dfx_generated/intermediary/intermediary.did';
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/cycles/test/tests.ts:6]: done

File: tests/property/ic_api/cycles/test/tests.ts (hunk 13 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/test/tests.ts:99]: sure

File: tests/property/ic_api/cycles/src/cycles/index.ts [hunk 15]
```diff
@@ -3,57 +3,55 @@ import {
...
-        //         `msgCyclesAvailable() !== 0. Actual value: ${remainingCycles}. Initial msgCyclesAvailable result: ${msgCyclesAvailableResult}`
-        //     );
-        // }
+    @update([IDL.Nat64], CyclesResult)
+    receiveCyclesByChunk(numChunks: bigint): CyclesResult {
+        return acceptCycles(undefined, numChunks);
     }
 }
 
 function acceptCycles(
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/cycles/src/cycles/index.ts:33]: done

File: tests/property/ic_api/cycles/src/cycles/index.ts [hunk 16]
```diff
@@ -3,57 +3,55 @@ import {
...
 ): CyclesResult {
     const startingCanisterBalance = canisterBalance();
     const initialAvailable = msgCyclesAvailable();
-    const accepted = msgCyclesAccept(getAcceptAmount(initialAvailable));
+
+    const effectiveNumChunks = numChunks ?? 1n;
+
+    const requestedAmount = receiveAmount ?? initialAvailable;
+
+    const chunkSize =
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/cycles/src/cycles/index.ts:44]: done

File: tests/property/ic_api/cycles/src/intermediary/index.ts [hunk 17]
```diff
@@ -1,52 +1,68 @@
...
+    @update([IDL.Nat64, IDL.Nat64], CyclesResult)
+    async sendVariableCycles(
+        amountToSend: bigint,
+        amountToAccept: bigint
+    ): Promise<CyclesResult> {
+        const result = await call(
+            this.cyclesPrincipal,
+            'receiveVariableCycles',
+            {
+                returnIdlType: CyclesResult,
```
[the remark this answers is not in the document]
Comment (bdemann) [reply tests/property/ic_api/cycles/src/intermediary/index.ts:26]: done

File: tests/property/ic_api/cycles/src/cycles/index.ts [hunk 18]
```diff
@@ -62,7 +54,18 @@ function acceptCycles(
     };
 }
 
-function acceptCyclesRecursive(
+function acceptCyclesStrategy(
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/src/cycles/index.ts:57]: Can we just have two functions? `acceptCyclesAll` and `acceptCyclesChunk`. Seems simpler

File: tests/property/ic_api/cycles/src/cycles/index.ts [hunk 19]
```diff
@@ -36,18 +36,10 @@ function acceptCycles(
...
-    const requestedAmount = receiveAmount ?? initialAvailable;
-
-    const chunkSize =
-        requestedAmount > 0n
-            ? bigintMax(1n, requestedAmount / effectiveNumChunks)
-            : 0n;
-
-    const accepted = acceptCyclesRecursive(chunkSize, requestedAmount, 0n);
-
+    const accepted = acceptCyclesStrategy(
```
Comment (lastmjs) [inline tests/property/ic_api/cycles/src/cycles/index.ts:39]: right here just do a ternary and call `acceptCyclesAll` or `acceptCyclesChunk`

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: tests/property/ic_api/cycles/src/cycles/index.ts (hunk 18 above)
Comment (bdemann) [reply tests/property/ic_api/cycles/src/cycles/index.ts]: I am not convinced that this is either simpler or more declarative, but if you like it here it is

**Review by lastmjs — APPROVED** [review]

**Review by lastmjs — APPROVED** [review]

**Pull request merged**